plugin-server/src/main/ingestion-queues/analytics-events-ingestion-consumer.ts

```diff
@@ -78,50 +78,63 @@ export async function eachBatchIngestionWithOverflow(
             const seenKey = `${pluginEvent.team_id}:${pluginEvent.distinct_id}`
-            if (currentBatch.length === batchSize || seenIds.has(seenKey)) {
+
+            // Events with a null key should have been produced to the the
+            // KAFKA_EVENTS_PLUGIN_INGESTION_OVERFLOW topic, so we shouldn't see them here as this consumer's
+            // topic is set to KAFKA_EVENTS_PLUGIN_INGESTION. However, there could be some lingering events
+            // from before the new *_OVERFLOW topic was initialized. Any events with a null key or that
+            // exceed capacity are redirected to the *_OVERFLOW topic.
+            if (message.key == null || ConfiguredLimiter.consume(seenKey, 1) === false) {
+                // Set message key to be null so we know to send it to overflow topic.
+                // We don't want to do it here to preserve the kafka offset handling
+                message.key = null
+            }
+
+            if (currentBatch.length === batchSize || (message.key != null && seenIds.has(seenKey))) {
                 seenIds.clear()
                 batches.push(currentBatch)
                 currentBatch = []
             }
-            seenIds.add(seenKey)
+
             currentBatch.push(message)
+
+            if (message.key != null) {
+                seenIds.add(seenKey)
+            }
         }
+
         if (currentBatch) {
             batches.push(currentBatch)
         }
+
         return batches
     }
 
     await eachBatch(payload, queue, eachMessageIngestionWithOverflow, groupIntoBatchesIngestion, 'ingestion')
 }
 
 export async function eachMessageIngestionWithOverflow(message: KafkaMessage, queue: IngestionConsumer): Promise<void> {
-    const event = formPipelineEvent(message)
-
-    if (
-        // Events with a null key are produced to the the KAFKA_EVENTS_PLUGIN_INGESTION_OVERFLOW topic, so we would
-        // never see them here, as this consumer's topic is set to KAFKA_EVENTS_PLUGIN_INGESTION.
-        // However, there could be some lingering events from before the new *_OVERFLOW topic was initialized.
-        // As we cannot accurately estimate capacity usage for those events, we will process them normally.
-        // Once lingering events are cleared up, this check is not be needed anymore except for type-checking.
-        message.key != null &&
-        ConfiguredLimiter.consume(message.key.toString(), 1) === false
-    ) {
-        if (event.team_id && WarningLimiter.consume(message.key.toString(), 1)) {
-            captureIngestionWarning(queue.pluginsServer.hub.db, event.team_id, 'ingestion_capacity_overflow', {
-                overflowDistinctId: event.distinct_id,
+    // Events are marked to have a null key during batch break-up if they should go to the *_OVERFLOW topic.
+    // So we do not ingest them here.
+    if (message.key == null) {
+        const { team_id: teamId, distinct_id: distinctId } = JSON.parse(message.value!.toString())
```

Comment:
Not entirely sure about this parsing thing. Didn't want to parse the entire event again.

Reply:
here's an idea: ingestion warnings don't need to be there immediately, so how about we just add the ingestion warning in from the overflow topic consumption?

Reply:
I also suspect this doesn't work based on looking at `formPipelineEvent`

Reply:
yes!

Reply:
Did this change and added some tests too.